src/lib/structures/Argument.ts

```diff
@@ -116,10 +117,106 @@ export abstract class Argument<T = unknown> extends BaseAliasPiece implements IA
 	}
 }
 
+/**
+ * The extended argument class. This class is abstract and is to be extended by subclasses which
+ * will implement the [[ExtendedArgument#handle]] method.
+ * Much like the [[Argument]] class, this class handles parsing user-specified command arguments
+ * into typed command parameters. However, this class can be used to expand upon an existing
+ * argument in order to process its transformed value rather than just the argument string.
+ *
+ * @example
+ * ```typescript
+ * // TypeScript:
+ * import { ApplyOptions } from '@sapphire/decorators';
+ * import { ArgumentResult, ExtendedArgument, ExtendedArgumentContext, ExtendedArgumentOptions } from '@sapphire/framework';
+ * import type { Channel, TextChannel } from 'discord.js';
+ *
+ * // Just like with `Argument`, you can use `export default` or `export =` too.
+ * @ApplyOptions<ExtendedArgumentOptions>({
+ *   name: 'textChannel',
+ *   baseArgument: 'channel'
+ * })
+ * export class TextChannelArgument extends ExtendedArgument<'channel', TextChannel> {
+ *   public handle(parsed: Channel, { argument }: ExtendedArgumentContext): ArgumentResult<TextChannel> {
+ *     return parsed.type === 'text'
+ *       ? this.ok(parsed as TextChannel)
+ *       : this.error(argument, 'ArgumentTextChannelInvalidTextChannel', 'The argument did not resolve to a text channel.');
+ *   }
+ * }
+ * ```
+ *
+ * @example
+ * ```javascript
+ * // JavaScript:
+ * const { ExtendedArgument } = require('@sapphire/framework');
+ *
+ * module.exports = class TextChannelArgument extends ExtendedArgument {
+ *   constructor(context) {
+ *     super(context, { name: 'textChannel', baseArgument: 'channel' });
+ *   }
+ *
+ *   handle(parsed, { argument }) {
+ *     return parsed.type === 'text'
+ *       ? this.ok(parsed)
+ *       : this.error(argument, 'ArgumentTextChannelInvalidTextChannel', 'The argument did not resolve to a text channel/');
+ *   }
+ * }
+ * ```
+ */
+export abstract class ExtendedArgument<K extends keyof ArgType, T> extends Argument<T> {
+	public baseArgument: K;
+
+	public constructor(context: PieceContext, options: ExtendedArgumentOptions<K>) {
+		super(context, options);
+		this.baseArgument = options.baseArgument;
+	}
+
+	/**
+	 * Represents the underlying argument that transforms the raw argument
+	 * into the value used to compute the extended argument's value.
+	 */
+	public get base(): IArgument<ArgType[K]> {
+		return this.client.arguments.get(this.baseArgument) as IArgument<ArgType[K]>;
+	}
+
+	public async run(argument: string, context: ArgumentContext): AsyncArgumentResult<T> {
+		const result = await this.base.run(argument, context);
+		return isOk(result)
+			? // The base argument resolved; handle it and use it to compute the extended argument.
+			  this.handle(result.value, { ...context, argument })
+			: // The base argument failed to resolve.
+			  this.error(argument, 'ExtendedArgumentInvalidArgument', `The argument did not resolve to a ${this.base.name}.`);
```

Comment:
I think it should return `result` as it's already an erroneous one.

Reply:
Should be resolved as of 8bc4a3d.